### PR TITLE
Allow specifying a keyfile with cert-based authentication

### PR DESCRIPTION
**Pull request by andugga**
Add certificate based authentication

### CODE REVIEW

File: restler/engine/transport_layer/messaging.py [hunk 1]
```diff
@@ -77,7 +80,7 @@ def _get_method_from_message(self, message):
         method_name = message[0:end_of_method_idx]
         return method_name
 
-    def sendRecv(self, message, req_timeout_sec):
+    def sendRecv(self, message, req_timeout_sec=600):
```
Comment (marina-p) [inline restler/engine/transport_layer/messaging.py:83]: >600 [](http://example.com/codeflow?start=48&length=3)

please sync to the latest in main and rebase - this should not be part of the changes

File: docs/user-guide/Authentication.md [hunk 2]
```diff
@@ -30,4 +32,9 @@ Note: in the above example, there are two different applications.  This is only
...
 **Token values in logs**
 
-RESTler has logic to prevent token values from being written to the network logs.  It is recommended to check the RESTler network logs and make sure that the token values are, indeed,  successfully omitted from the logs. 
\ No newline at end of file
+RESTler has logic to prevent token values from being written to the network logs.  It is recommended to check the RESTler network logs and make sure that the token values are, indeed,  successfully omitted from the logs. 
+                
+**Certificate based authentication**
+                
+Certificate and corresponding keys can be used as an authentication machanism. To use this add a key value pair in the settings.json file e.g. 
+                "client_certificate_key_path":"C:\\Users\\<folder></folder>\\privateKey.txt"
```
Comment (marina-p) [inline docs/user-guide/Authentication.md:40]: >client_certificate_key_path [](http://example.com/codeflow?start=17&length=27)

This setting should be added to SettingsFile.md, right below client_certificate_path.  

It should say something like:

Path to the keyfile that is needed to load the certificate.

If both the keyfile and certificate path are valid, RESTler will attempt to use it during the SSL handshake.

File: docs/user-guide/Authentication.md [hunk 3]
```diff
@@ -1,6 +1,8 @@
 # Authentication
 
-RESTler supports token-based authentication.  
+RESTler supports token-based and cert based authentication.
```
Comment (marina-p) [inline docs/user-guide/Authentication.md:3]: >cert [](http://example.com/codeflow?start=33&length=4)

certificate-based

File: docs/user-guide/Authentication.md (hunk 2 above)
Comment (marina-p) [inline docs/user-guide/Authentication.md:40]: You can just say 'See the SettingsFile.md for the settings that should be used to specify certificates.

File: restler/engine/transport_layer/messaging.py [hunk 4]
```diff
@@ -53,12 +53,15 @@ def __init__(self, connection_settings):
                     context = ssl._create_unverified_context()
                 else:
                     context = ssl.create_default_context()
-                if Settings().client_certificate_path:
+                if Settings().client_certificate_path and Settings().client_certificate_key_path:
                     context.load_cert_chain(
```
Comment (marina-p) [inline restler/engine/transport_layer/messaging.py:57]: If you add this 'and', the code path without a key file is broken.  I believe if you just remove the 'and', the previous code path will continue to work.

**Review by marina-p — CHANGES_REQUESTED** [review]
:clock1:

**Review by andugga — COMMENTED** [review]
Addressed the comments from the review

File: restler/engine/transport_layer/messaging.py [hunk 5]
```diff
@@ -55,10 +55,13 @@ def __init__(self, connection_settings):
                     context = ssl.create_default_context()
                 if Settings().client_certificate_path:
                     context.load_cert_chain(
-                        certfile=Settings().client_certificate_path
+                        certfile = Settings().client_certificate_path,
+                        keyfile = Settings().client_certificate_key_path if Settings().client_certificate_key_path else None,
```
Comment (marina-p) [inline restler/engine/transport_layer/messaging.py:59]: >if Settings().client_certificate_key_path else None, [](http://example.com/codeflow?start=73&length=52)

The default setting is already None, so you can omit this if

File: docs/user-guide/Authentication.md [hunk 6]
```diff
@@ -30,4 +32,8 @@ Note: in the above example, there are two different applications.  This is only
 
 **Token values in logs**
 
-RESTler has logic to prevent token values from being written to the network logs.  It is recommended to check the RESTler network logs and make sure that the token values are, indeed,  successfully omitted from the logs. 
\ No newline at end of file
+RESTler has logic to prevent token values from being written to the network logs.  It is recommended to check the RESTler network logs and make sure that the token values are, indeed,  successfully omitted from the logs. 
+                
+**Certificate based authentication**
+                
+Certificate and corresponding keys can be used as an authentication machanism. Path to the keyfile that is needed to load the certificate. To use this add a key value pair in the settings.json. See the SettingsFile.md for the settings that should be used to specify a certificate. If both the keyfile and certificate path are valid, RESTler will attempt to use it during the SSL handshake.
```
Comment (marina-p) [inline docs/user-guide/Authentication.md:39]: >machanism [](http://example.com/codeflow?start=68&length=9)

mechanism

Comment (marina-p) [inline docs/user-guide/Authentication.md:39]: >Path to the keyfile that is needed to load the certificate. To use this add a key value pair in the settings.json. [](http://example.com/codeflow?start=79&length=114)

These two sentences can be removed.

Comment (marina-p) [inline docs/user-guide/Authentication.md:39]: >Certificate [](http://example.com/codeflow?start=0&length=11)

A certificate

File: restler/restler_settings.py [hunk 7]
```diff
@@ -364,6 +364,8 @@ def convert_wildcards_to_regex(str_value):
 
         ## Path to Client Cert for Certificate Based Authentication
         self._client_certificate_path = SettingsArg('client_certificate_path', str, None, user_args)
+        ## Path to Client Cert Key for Certificate Based Authentication
+        self._client_certificate_key_path = SettingsArg('client_certificate_key_path', str, None, user_args)
```
Comment (marina-p) [inline restler/restler_settings.py:368]: >_client_certificate_key_path [](http://example.com/codeflow?start=13&length=28)

Please document this in SettingsFile.md (it is not generated automatically)

**Review by marina-p — APPROVED** [review]
:shipit:

**Pull request merged**